From 1159bd45fbcfea2678b44796557c98ce427eafd4 Mon Sep 17 00:00:00 2001
From: Xianzhong <xianzhong.li@nxp.com>
Date: Thu, 13 Jul 2017 10:28:14 +0800
Subject: [PATCH 5217/5242] MGS-3083 [#imx-662] fix kernel build error with
 kasan config

commit  84dc4728736c6b6d70be0e0952100ac92e436649 from
https://source.codeaurora.org/external/imx/linux-imx.git

hal/kernel/gc_hal_kernel_command.c: In function gckCOMMAND_Commit:
hal/kernel/gc_hal_kernel_command.c:2718:1:
error: the frame size of 2288 bytes is larger than 2048 bytes [-Werror=frame-larger-than=]

removed _commandBufferObject instrance, use kmalloc to avoid kasan error.

Date: Jul 13, 2017
Signed-off-by: Xianzhong <xianzhong.li@nxp.com>
(cherry picked from commit d8d57e43a24bd02b375d7c6bd633ed236db19337)
(cherry picked from commit 484e3073adf47d27581482f268b01fe5fe219d18)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../mxc/gpu-viv/hal/kernel/gc_hal_kernel_command.c |   12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel_command.c b/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel_command.c
index 528860c..2db8e4d 100644
--- a/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel_command.c
+++ b/drivers/mxc/gpu-viv/hal/kernel/gc_hal_kernel_command.c
@@ -1790,7 +1790,6 @@
 
 #if !gcdNULL_DRIVER
     gcsCONTEXT_PTR contextBuffer;
-    struct _gcoCMDBUF _commandBufferObject;
     gctPHYS_ADDR_T commandBufferPhysical;
     gctUINT8_PTR commandBufferLogical = gcvNULL;
     gctUINT32 commandBufferAddress = 0;
@@ -1904,7 +1903,8 @@
 #else
     if (needCopy)
     {
-        commandBufferObject = &_commandBufferObject;
+        gcmkONERROR(gckOS_Allocate(Command->os, gcmSIZEOF(struct _gcoCMDBUF), &pointer));
+        commandBufferObject = pointer;
 
         gcmkONERROR(gckOS_CopyFromUserData(
             Command->os,
@@ -2587,6 +2587,10 @@
 
         commandBufferMapped = gcvFALSE;
     }
+    else if (needCopy)
+    {
+        gcmkONERROR(gckOS_Free(Command->os, commandBufferObject));
+    }
 
     /* Return status. */
     gcmkFOOTER();
@@ -2626,6 +2630,10 @@
             commandBufferObject
             ));
     }
+    else if (needCopy)
+    {
+        gcmkVERIFY_OK(gckOS_Free(Command->os, commandBufferObject));
+    }
 
     /* Return status. */
     gcmkFOOTER();
-- 
1.7.9.5

